statscache/producer.py

```diff
@@ -13,81 +14,94 @@
-        self.sister = statscache.utils.find_stats_consumer(self.hub)
+        # set up the cache and connect to the consumer
+        self.cache = []
+        statscache.utils.find_stats_consumer(self.hub).register(self)
 
-        self.plugins = statscache.utils.load_plugins(self.hub.config)
+        #self.plugins = statscache.utils.init_plugins(self.hub.config)
 
         uri = self.hub.config['statscache.sqlalchemy.uri']
         statscache.plugins.create_tables(uri)
 
         log.debug("%s initialized with %r plugins" % (
-            self.name, len(self.plugins)))
-        self.init_plugins()
-
-    def init_plugins(self):
-        """ Allow plugins to initialize themselves using the database """
+            type(self).__name__, len(self.plugin_classes)))
         session = self.make_session()
-        for plugin in self.plugins.values():
-            initialize = getattr(plugin, 'initialize', None)
-            if initialize is None:
-                continue
-            plugin.initialize(session)
-        session.commit()
+        self.plugins = []
+        for plugin_class in self.plugin_classes:
+            try:
+                plugin = plugin_class(self.frequency, self.hub.config)
+                initialize = getattr(plugin, 'initialize', None)
+                if initialize is not None:
+                    plugin.initialize(session)
+                    session.commit()
+                self.plugins.append(plugin)
+                log.info("Initialized plugin %r" % plugin.ident)
+            except Exception:
+                log.exception("Failed to initialize plugin %r" % plugin_class)
+                session.rollback()
 
     def make_session(self):
         """ Initiate database connection """
         uri = self.hub.config['statscache.sqlalchemy.uri']
         return statscache.plugins.init_model(uri)
 
+    def buffer(self, message):
+        self.cache.append(message)
+
     def poll(self):
         """
-        Empty the associated bucket and send to each plugin for processing.
+        Empty the cache and distribute the contents to each plugin for
+        processing.
         """
-        now = datetime.datetime.utcnow()
+        timestamp = datetime.datetime.utcnow() # moment cache was cleared
 
-        bucket = self.sister.buckets[self.name]
-        self.sister.buckets[self.name] = []
+        cache = self.cache
+        self.cache = []
 
-        n = len(bucket)
-        log.info("%s called with %i items in the bucket." % (self.name, n))
+        n = len(cache)
+        log.info("%s called with %i cached items." % (type(self).__name__, n))
 
-        for plugin in self.plugins.values():
-            log.info("  Calling %r" % plugin.name)
+        for plugin in self.plugins:
+            log.info("  Calling %r" % plugin.ident)
             session = self.make_session()
             try:
-                plugin.handle(session, bucket)
+                plugin.handle(session, timestamp, cache)
```

Comment:
Also, can we get rid of the ``timestamp`` in the plugin's ``handle`` call and signatures. The vision was for plugins to be stateless, and use the timestamp from fedmsg message.